### ISSUE TITLE
Change how config files are searched in upper directories

### ISSUE THREAD

**Issue by fvictorio** [NomicFoundation/hardhat#3477]
The way Hardhat currently searches for a config is:
- Hardhat searches for a `hardhat.config.ts` in the current directory, and then in the ancestor directories.
- If none is found, the search starts over from the current directory, but a `hardhat.config.js` file is searched instead.

This is problematic [in some scenarios](https://github.com/NomicFoundation/hardhat/issues/2669). Instead, we should search for any possible config file in the current directory, and then in the parent directory, and so on.

I believe this is technically a breaking change, so I'm going to mark it as so. But you could argue it's a bug fix. If a lot of people need this and we are not about to release a new major soon, I could consider doing it in v2, since I feel it's unlikely to break someone's workflow. But, [you know...](https://xkcd.com/1172/)

**Comment by neila** [comment]
I'm experiencing this precise issue right now, and would also love to see this search algorithm improved.
I am working with a hardhat with typescript [setup](https://github.com/neila/stakingpools/tree/main) while referring to [multiple submodules](https://github.com/neila/stakingpools/tree/main/lib) of smart contract contract libraries. 
However, most if not all of the contracts from the submodules are developed using the [javascript](https://github.com/lidofinance/lido-dao/) [setup](https://github.com/rocket-pool/rocketpool/) (which is fair given the adoption of typescript in smart contract development is a relatively recent scene). When I try to run some of the testing scripts in the submodules, they don't work, because instead of applying the actual `hardhat.config.js` files from each setup, it applies the `.ts` file from my main repository beforehand.

Since each project is well-contained, hardhat could just apply the closest config file (and by that I mean in each project's root) and this wouldn't be a problem. Given the contained nature of node/hardhat environment, I think it would make a lot more sense to use closest config file, regardless of js/ts, allowing for a much more modular dev experience.

**Comment by fvictorio** [comment]
Yeah, that makes sense. I initially marked this as a breaking change, but I would be fine with filing this under "well, it's actually a fix!", if this is a big issue for your project.
